ui/src/shared/components/TimeMachineQueryBuilder.tsx

```diff
@@ -56,25 +61,40 @@ class TimeMachineQueryBuilder extends PureComponent<Props, State> {
               {range(tagFiltersLength).map(i => (
                 <TagSelector key={i} index={i} />
               ))}
-              <Button
-                shape={ButtonShape.Square}
-                icon={IconFont.Plus}
-                onClick={onAddTagSelector}
-                customClass="query-builder--add-tag-selector"
-              />
+              {this.addButton}
             </div>
           </FancyScrollbar>
           <FunctionSelector />
         </div>
       </div>
     )
   }
+
+  private get addButton(): JSX.Element {
+    const {moreTags, onAddTagSelector} = this.props
+
+    if (!moreTags) {
+      return null
+    }
+
+    return (
+      <Button
+        shape={ButtonShape.Square}
+        icon={IconFont.Plus}
+        onClick={onAddTagSelector}
+        customClass="query-builder--add-tag-selector"
+      />
+    )
+  }
 }
 
 const mstp = (state: AppState): StateProps => {
   const tagFiltersLength = getActiveQuery(state).builderConfig.tags.length
+  const moreTags = !getActiveTimeMachine(state).queryBuilder.tags.find(
+    t => !t.keys.length && t.keysStatus === RemoteDataState.Done
```

Comment:
I think this will hide the add button when a user searches for tag keys in any `TagSelector` and the search returns no results.

I think that's fine. Maybe a small improvement would be to only perform this check on the last `queryBuilder.tags` object (which corresponds to the right-most `TagSelector`).

Reply:
fair enough